src/frontend/src/scheduler/local.rs

```diff
@@ -105,16 +153,61 @@ impl LocalQueryExecution {
         })
     }
 
-    fn convert_plan_node(&self, execution_plan_node: &ExecutionPlanNode) -> Result<PlanNodeProst> {
+    fn convert_plan_node(
+        &self,
+        execution_plan_node: &ExecutionPlanNode,
+        second_stage_plan: &mut Option<HashMap<StageId, PlanFragment>>,
+    ) -> Result<PlanNodeProst> {
         match execution_plan_node.plan_node_type {
             PlanNodeType::BatchExchange => {
-                Err(internal_error("Exchange not supported in local mode yet!"))
+                let exchange_from_stage_id = execution_plan_node
+                    .stage_id
+                    .expect("We expect stage id for Exchange Operator");
+                match second_stage_plan.as_mut() {
+                    Some(second_stage_plan) => {
+                        let second_stage_plan_fragment = second_stage_plan.remove(&exchange_from_stage_id).expect("We expect child stage fragment for Exchange Operator running in the frontend");
+                        let mut exchange_node =
+                            try_match_expand!(execution_plan_node.node.clone(), NodeBody::Exchange)?;
+                        let local_execute_plan = LocalExecutePlan {
+                            plan: Some(second_stage_plan_fragment),
+                            epoch: self.epoch.expect("Local execution mode has not acquired the epoch when generating the plan.")
+                        };
+                        exchange_node.sources.extend(self.front_env.worker_node_manager().list_worker_nodes().iter().enumerate().map(|(idx, worker_node)| {
+                            let exchange_source = ExchangeSource {
+                                task_output_id: Some(TaskOutputId {
+                                    task_id: Some(ProstTaskId {
+                                        // We remark that `RowSeqScanExecutor` relies on the task_id
+                                        // to differentiate which one is primary and
+                                        // should really read all the data.
+                                        task_id: idx as u32,
+                                        stage_id: 0,
+                                        query_id: "local execution mode".to_string(),
```

Comment:
The `query_id` and `stage_id` should inherit from stage graph, and it helps debugging.